### ISSUE TITLE
MSAA Changes colors and adds borders around sprites

### ISSUE THREAD

**Issue by TheNeikos** [bevyengine/bevy#419]
When setting a `Msaa` level above 1 sample the color changes:

Msaa 1:
![image](https://user-images.githubusercontent.com/1631166/91944210-3d6a4380-eced-11ea-92c1-1423182dbc11.png)

Msaa 2:
![image](https://user-images.githubusercontent.com/1631166/91944238-4bb85f80-eced-11ea-8ccb-c8ae77e631df.png)

Msaa 4:
![image](https://user-images.githubusercontent.com/1631166/91944286-5a9f1200-eced-11ea-8421-3e825da6b910.png)

Code:

```rust
use bevy::prelude::*;
use bevy::render::pass::ClearColor;

fn main() {
    env_logger::init();
    App::build()
        .add_resource(Msaa { samples: 4 })
        .add_resource(ClearColor(Color::rgb(0.5, 0.5, 0.0)))
        .add_default_plugins()
        .add_startup_system(setup.system())
        .run();
}

struct Square(u64);

fn setup(
    mut commands: Commands,
    mut materials: ResMut<Assets<ColorMaterial>>,
    asset_server: Res<AssetServer>,
) {
    commands
        .spawn(Camera2dComponents::default())
        .spawn(UiCameraComponents::default())
        .spawn((Square(1), Translation::new(0., 20., 0.)))
        .with_children(|parent| {
            parent.spawn(SpriteComponents {
                material: materials.add(Color::rgb(1.0, 0.2, 0.2).into()),
                sprite: Sprite {
                    size: Vec2::new(25., 25.),
                },
                translation: Translation::new(0., 0., 0.),
                ..Default::default()
            });
        });
}

```

My platform is Linux with an NVIDIA Corporation GP107GLM [Quadro P620] GPU

**Comment by julhe** [comment]
Hmm, I can't reproduce this under Windows 10 x64 with a AMD R9 390. 

Could you test this https://github.com/gfx-rs/wgpu-rs/tree/master/examples/msaa-line? Change the MSAA level with the left and right arrow key and see if the problem is also present there.

**Comment by TheNeikos** [comment]
@julhe No, it doesn't happen there

**Comment by TheNeikos** [comment]
Another test on Win10 with an GTX 1060:

Msaa 1:
![image](https://user-images.githubusercontent.com/1631166/92990525-b5462400-f4cc-11ea-8870-bf20ee9c83fd.png)

Msaa 4:
![image](https://user-images.githubusercontent.com/1631166/92990529-c000b900-f4cc-11ea-92b7-f2b8d6c33fcb.png)

Msaa 8:
![image](https://user-images.githubusercontent.com/1631166/92990532-cb53e480-f4cc-11ea-9c67-c0dcf53c7945.png)

**Comment by cart** [comment]
As @StarArawn mentioned on discord, the weird multisampling behavior might be related to: https://github.com/gfx-rs/wgpu/issues/725. I think we should treat this issue as scoped to that problem.

A little bit of blurriness is to be expected with MSAA. It makes jagged edges crisper by taking multiple samples and blending them together (which is the literal definition of a blur operation). This makes jagged edges crisper, but it can make already straight lines a little bit blurry.

I do think we have _another_ issue though. I also get the same "incorrect color" results on my computer (linux, proprietary nvidia drivers). I'm pretty sure the incorrect "base" colors are related to bevy rendering with the srgb color space. I have a feeling we aren't doing a proper conversion here.  The math certainly lines up with improper color handling.

Desired linear color value: **0.5**
SRGB to Linear Conversion: ((**0.5** + 0.055) / 1.055) ^ 2.4 =  0.214
Rendering with Color::rgb(0.214, 0.214, 0.0) and color picking = rgb(128, 128, 0)

I'm pretty sure we're locked into an SRGB swap chain, so we might need to do these conversions internally.

**Comment by cart** [comment]
As a followup, loaded images do use the correct colors (and also use the SRGB image type), so I think the fix here is it convert the `Color` type before uploading to the shader.

**Comment by TheNeikos** [comment]
@cart I understand that some things might get a bit more blurry, but it also adds a discolored edge to everything which it imo should not do

For example, in my game this is what it looks with Msaa 8:
![image](https://user-images.githubusercontent.com/1631166/93016260-6de49400-f5af-11ea-97bb-e9365f3a4521.png)

and without:
![image](https://user-images.githubusercontent.com/1631166/93016269-7b9a1980-f5af-11ea-9f1c-202427121a15.png)

**Comment by cart** [comment]
@TheNeikos hmm yeah thats not desirable. I'm honestly uncertain if thats a bevy issue, a wgpu issue, or a general MSAA issue.

**Comment by cwfitzgerald** [comment]
Fully agree that this is just https://github.com/gfx-rs/wgpu/issues/725. It manifests exactly as the photos show; borked srgb conversion of the background plus bad MSAA. It's been a long-standing issue with intel (at least on linux), and just needs to be worked around at some layer.

**Comment by rparrett** [comment]
https://github.com/gfx-rs/wgpu/issues/725 is apparently fixed since wgpu 0.11.

Anyone affected still able to reproduce with bevy main?

**Comment by cart** [comment]
I think its probably safe to close this for now and then re-open if this resurfaces.
